go/jsonclient/client.go

```diff
@@ -0,0 +1,191 @@
+	httpClient *http.Client          // used to interact with the server via HTTP
+	Verifier   *ct.SignatureVerifier // nil if no public key for server available
+}
+
+// New constructs a new JSONClient instance, for the given base URI, using the
+// given http.Client object (if provided).  If a PEM encoded public key is
+// provided, the JSONClient instance will be able to check signatures on signed
+// responses from the server.
+func New(uri string, hc *http.Client, pemEncodedKey string) (*JSONClient, error) {
+	var verifier *ct.SignatureVerifier
+	if pemEncodedKey != "" {
+		pubkey, _, rest, err := ct.PublicKeyFromPEM([]byte(pemEncodedKey))
+		if err != nil {
+			return nil, err
+		}
+		if len(rest) > 0 {
+			return nil, errors.New("extra data found after PEM key decoded")
+		}
+
+		verifier, err = ct.NewSignatureVerifier(pubkey)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if hc == nil {
+		hc = new(http.Client)
+	}
+	return &JSONClient{uri: strings.TrimRight(uri, "/"), httpClient: hc, Verifier: verifier}, nil
+}
+
+// GetAndParse makes a HTTP GET call to the given path, and attempt to parse
+// the response as a JSON representation of the rsp structure.  The provided
+// context is used to control the HTTP call.
+func (c *JSONClient) GetAndParse(ctx context.Context, path string, params map[string]string, rsp interface{}) (*http.Response, error) {
+	if ctx == nil {
+		return nil, errors.New("Context required")
+	}
+	// Build a GET request with URL-encoded parameters.
+	vals := url.Values{}
+	for k, v := range params {
+		vals.Add(k, v)
+	}
+	fullURI := fmt.Sprintf("%s%s?%s", c.uri, path, vals.Encode())
+	httpReq, err := http.NewRequest(http.MethodGet, fullURI, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	httpRsp, err := ctxhttp.Do(ctx, c.httpClient, httpReq)
+	if err != nil {
+		return nil, err
+	}
+	// Make sure everything is read, so http.Client can reuse the connection.
+	defer httpRsp.Body.Close()
+	defer ioutil.ReadAll(httpRsp.Body)
+
+	if httpRsp.StatusCode != http.StatusOK {
+		return httpRsp, fmt.Errorf("got HTTP Status %s", httpRsp.Status)
+	}
+
+	if err := json.NewDecoder(httpRsp.Body).Decode(rsp); err != nil {
+		return httpRsp, err
+	}
+
+	return httpRsp, nil
+}
+
+// PostAndParse makes a HTTP POST call to the given path, including the request
+// parameters, and attempt to parse the response as a JSON representation of
+// the rsp structure.  The provided context is used the control the HTTP call.
+func (c *JSONClient) PostAndParse(ctx context.Context, path string, req interface{}, rsp interface{}) (*http.Response, error) {
+	if ctx == nil {
+		return nil, errors.New("Context required")
+	}
+	// Build a POST request with JSON body.
+	postBody, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	fullURI := fmt.Sprintf("%s%s", c.uri, path)
+	httpReq, err := http.NewRequest(http.MethodPost, fullURI, bytes.NewReader(postBody))
```

Comment:
I'd go for some briefer names, eg. 'req' and 'resp' in place of httpReq, httpRsp
Go readability folk are keen on brevity. Particularly where the point of declaration is close to the final point of use.
You might then use the longer, more readable names for the function arguments.